File: problems/959-regions-cut-by-slashes/dsu.cpp

/**
 * 并查集
 * 
 * 时间：64ms
 */

#include <vector>
#include <string>
using namespace std;

class UnionFind {
public:
    UnionFind(int size) {
        for (int i = 0; i < size; ++i) {
            parent.push_back(i);
        }
        rank = vector<int>(size, 1);
    }

    void merge(int i, int j) {
        int ii = getRoot(i);
        int jj = getRoot(j);
        if (ii == jj) return;

        if (rank[ii] <= rank[jj]) {
            parent[ii] = jj;
            if (rank[ii] == rank[jj]) {
                ++rank[jj];
            }
        } else {
            parent[jj] = ii;
        }
    }

    int getRoot(int x) {
        if (parent[x] == x) return x;
        return parent[x] = getRoot(parent[x]); // 路径压缩
    }
    
private:
    vector<int> parent;
    vector<int> rank;
};

const vector<vector<int>> DIRS = {
    {-1, 0, 0, 2}, // { 格子x偏移，格子y偏移，小格子偏移，相邻小格子偏移 }
    {1, 0, 2, 0},
    {0, -1, 3, 1},
    {0, 1, 1, 3}
};

class Solution {
public:
    int regionsBySlashes(vector<string>& grid) {
        const int n = grid.size();
        UnionFind uf(n * n * 4); // 每个格子分为4个小格子

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                char ch = grid[i][j];
                int k = 4 * (i * n + j); // 小格子起始下标

                // 格子内部的连通性
                if (ch != '\\') {
                    uf.merge(k, k + 3);
                    uf.merge(k + 1, k + 2);
                }
                if (ch != '/') {
                    uf.merge(k, k + 1);
                    uf.merge(k + 2, k + 3);
                }

                // 当前格子与相邻格子的连通性
                for (auto dir : DIRS) {
                    int ii = i + dir[0];
                    int jj = j + dir[1];
                    if (ii >= 0 && ii < n && jj >= 0 && jj < n) {
                        uf.merge(k + dir[2], 4 * (ii * n + jj) + dir[3]);
                    }
                }
            }
        }

        unordered_set<int> roots;
        for (int i = n * n * 4 - 1; i >= 0; --i) {
            roots.insert(uf.getRoot(i));
        }
        return roots.size();
    }
};

File: weekly-contest/weekly-contest-181/3-check-if-there-is-a-valid-path-in-a-grid/dsu.js
/**
 * 并查集
 */
var hasValidPath = function(grid) {
  const m = grid.length, n = grid[0].length;
  const DIRS = [[-1, 0], [0, -1], [1, 0], [0, 1]];
  const uf = new UnionFind(m * n);

  function getKey(i, j) {
    return i * n + j;
  }

  function isValid(ni, nj) {
    return ni >= 0 && ni < m && nj >= 0 && nj < n;
  }

  function isConnected(r1, c1, r2, c2, dir) {
    const p = [grid[r1][c1], grid[r2][c2]];
    let pairs;
    if (dir === 0) { // 上
      pairs = [[2,5,6], [2,3,4]];
    } else if (dir === 1) { // 左
      pairs = [[1,3,5], [1,4,6]];
    } else if (dir === 2) { // 下
      pairs = [[2,3,4], [2,5,6]];
    } else { // 右
      pairs = [[1,4,6], [1,3,5]];
    }
    return pairs.every((_, i) => pairs[i].includes(p[i]));
  }

  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      const k1 = getKey(i, j);
      // 看相邻的4个方向，是否可以连通
      for (let k = 0; k < 4; ++k) {
        const ni = i + DIRS[k][0];
        const nj = j + DIRS[k][1];
        if (isValid(ni, nj) && isConnected(i, j, ni, nj, k)) {
          const k2 = getKey(ni, nj);
          uf.union(k1, k2);
        }
      }
    }
  }

  return uf.find(0, m * n - 1);
};

class UnionFind {
  constructor (length) {
    this.father = []
    this.rank = []

    for (let i = 0; i < length; ++i) {
      this.father[i] = i;
      this.rank[i] = 1;
    }
  }

  getRoot (x) {
    if (x === this.father[x]) return x
    return this.father[x] = this.getRoot(this.father[x]) // 路径压缩
  }

  find (x, y) {
    return this.getRoot(x) === this.getRoot(y)
  }

  union (x, y) {
    let xx = this.getRoot(x)
    let yy = this.getRoot(y)
    if (xx === yy) return false

    // 建立连通时，让rank更大的作为“根”
    if (this.rank[xx] <= this.rank[yy]) {
      this.father[xx] = yy;
      if (this.rank[xx] == this.rank[yy]) {
        ++this.rank[yy];
      }
    } else {
      this.father[yy] = xx;
    }
    return true
  }
}

[
  [[2,4,3],[6,5,2]],
  [[1,2,1],[1,2,1]],
  [[1,1,2]],
  [[1,1,1,1,1,1,3]],
  [[2],[2],[2],[2],[2],[2],[6]]
].forEach(A => {
  console.log(hasValidPath(A))
})